tests/test_core.py

```diff
@@ -7350,6 +7350,93 @@ def source_map_file_loc(name):
     else:
       assert seen_lines.issuperset([6, 7, 11, 12]), seen_lines
 
+  @no_wasm2js('TODO: source maps in wasm2js')
+  @no_fastcomp('DWARF is only supported in upstream')
+  def test_dwarf(self):
+    self.emcc_args.append('-gforce_dwarf')
+    self.emcc_args.remove('-Werror') # ignore warning on force-dwarf
+
+    create_test_file('src.cpp', '''
+      #include <stdio.h>
+      #include <assert.h>
+
+      __attribute__((noinline)) int foo() {
+        printf("hi"); // line 6
+        return 1; // line 7
+      }
+
+      int main() {
+        printf("%d", foo()); // line 11
+        return 0; // line 12
+      }
+    ''')
+
+    js_filename = 'a.out.js'
+    wasm_filename = 'a.out.wasm'
+
+    Building.emcc('src.cpp',
+                  self.serialize_settings() + self.emcc_args,
+                  js_filename)
+
+    LLVM_DWARFDUMP = os.path.join(LLVM_ROOT, 'llvm-dwarfdump')
+    out = run_process([LLVM_DWARFDUMP, wasm_filename, '-all'], stdout=PIPE).stdout
+
+    # parse the sections
+    sections = {}
+    curr_section_name = ''
+    curr_section_body = ''
+
+    def add_section():
+      if curr_section_name:
+        sections[curr_section_name] = curr_section_body
+
+    for line in out.splitlines():
+      if ' contents:' in line:
+        # a new section, a line like ".debug_str contents:"
+        add_section()
+        curr_section_name = line.split(' ')[0]
+        curr_section_body = ''
+      else:
+        # possibly a line in a section
+        if curr_section_name:
+          curr_section_body += line + '\n'
+    add_section()
+
+    # make sure the right sections exist
+    self.assertIn('.debug_abbrev', sections)
+    self.assertIn('.debug_info', sections)
+    self.assertIn('.debug_line', sections)
+    self.assertIn('.debug_str', sections)
+    self.assertIn('.debug_ranges', sections)
+
+    # verify some content in the sections
+    self.assertIn('"src.cpp"', sections['.debug_info'])
```

Comment:
do we want to assert anything about the path?

Reply:
Not sure what you mean by "path" here? Do you mean to check the DIE that `"src.cpp"` string is in, or something like that?

Reply:
Or do you mean the full path to `src.cpp`? It's encoded as a relative address, that's the whole string.

Reply:
yeah that's what I meant. it makes sense this way.